// npm/module_vk.js
vk = (function(){
if(typeof require == "function"){
	var request = require('request');
	var requestSync = require('sync-request');
	var browser = false;
}else{
	console.info("I am in browser");
	var browser = true;
}
var vk = {
request_prompt:0,
inbrowser:browser,
request:request,
requestSync:requestSync,
api_url:"https://api.vk.com/method/",
default_version:"5.45",
access_token:"",
debag:1,
uid:{me:61351294,bot:302799706},
default_callback:function(a){if(vk.debag)console.log(a);},
upload:function(first_method,save_method,data,cb){
	//ядро загрузки файлов
	//first_method - метод для получения сервера
	//save_method - метод для сохранения файла
	//data - данные:
	//data["files"] должен содержать файлы для загрузки
	if(vk.debag)console.log("Получаем сервер для загрузки файлов...");
	var _data = vk.other.collect(data,{files:""});
	first_method(_data,function(s){
		if(vk.debag){
			console.log("Результаты получения сервера для загрузки...",s);
			console.log("Загрузка файлов...");
		}
		vk.sreq.post(s.response.upload_url,data["files"],function(a){
				if(vk.debag)console.log("Файлы загружены, сохраняю их...",a);
				var save_data = vk.other.collect(s.response,_data,a);
				save_method(save_data,cb);
		})
	});
},
other:{
	collect: function() {
	var ret = {};
	var len = arguments.length;
	for (var i=0; i<len; i++) {
	 for (p in arguments[i]) {
		if (arguments[i].hasOwnProperty(p)) {
		  ret[p] = arguments[i][p];
		}
	 }
	}
	return ret;
	}
},
sreq:{
		get:function(u,c,e){
			if(typeof request == "undefined")return console.error("request is not defined");
			if(!e)e = function(b){console.log(b)};
			if(!c)c = function(b){console.log(b)};
			return  request.get(u, function optionalCallback(err, httpResponse, body) {
				if(err)return e(err);
				if(global.gc)global.gc();
				var t;
				try{
					t = JSON.parse(body);
				}catch(e){
					t = body;
				}
				return c(t);
			});	
		},
		getSync:function(u){
			if(typeof requestSync == "undefined")return console.error("requestSync is not defined");
			var res = requestSync('GET', u);
			if(global.gc)global.gc();
			try{
				return JSON.parse(res.getBody('utf8'));
			}catch(e){
				return res.getBody('utf8');
			}
		},
		postSync:function(u,d){
			if(typeof requestSync == "undefined")return console.error("requestSync is not defined");
			if(!d)d = {};
			var res = requestSync('POST', u, {qs: d});
			if(global.gc)global.gc();
			try{
				return JSON.parse(res.getBody('utf8'));
			}catch(e){
				return res.getBody('utf8');
			}
		},
		post:function(u,d,c,e){
			if(typeof request == "undefined")return console.error("request is not defined");
			if(!e)e = function(b){console.log(b)};
			if(!c)c = function(b){console.log(b)};
			for(var i in d){
				if(typeof d[i] == "undefined")delete d[i];
			}
			return request.post({url:u, formData: d}, function optionalCallback(err, httpResponse, body) {
			if(err)return e(err);
			if(global.gc)global.gc();
			var t;
			try{
				t = JSON.parse(body);
			}catch(e){
				t = body;
			}
			return c(t);
			});
		},
		jsonp:function(url,d, onSuccess, onError){
			if(!onError)onError = function(a){console.log(a)};
			if(!onSuccess)onSuccess = function(a){console.log(a)};
			var scriptOk = false;
			var callbackName = 'f'+String(Math.random()).slice(2);
			url += ~url.indexOf('?') ? '&' : '?';
			url += 'callback=CallbackRegistry.'+callbackName;
			for(var n in d)url += '&'+n+'='+encodeURIComponent(d[n]);
			CallbackRegistry[callbackName] = function(data) {       
			scriptOk = true;
			try{
				document.getElementById(callbackName).remove();
			}catch(e){}
				delete CallbackRegistry[callbackName];
				onSuccess(data);
			};
			function checkCallback() {      
				if (scriptOk) return;
				delete CallbackRegistry[callbackName]; 
				delete document.getElementById(callbackName).remove();
				onError(url);
			}
			var script = document.createElement('script');
			script.onreadystatechange = function(){  
				if (this.readyState == 'complete' || this.readyState == 'loaded'){   
					this.onreadystatechange = null;
					setTimeout(checkCallback, 0);
				}
			}
			script.onload = script.onerror = checkCallback;
			script.src = url;
			script.id = callbackName;
			document.body.appendChild(script);
			}		
	},
permissions:[
	["notify",1,"Пользователь разрешил отправлять ему уведомления (для flash/iframe-приложений)."],
	["friends",2,"Доступ к друзьям."],
	["photos",4,"Доступ к фотографиям."],
	["audio",8,"Доступ к аудиозаписям."],
	["video",16,"Доступ к видеозаписям."],
	["docs",131072,"Доступ к документам."],
	["notes",2048,"Доступ к заметкам пользователя."],
	["pages",128,"Доступ к wiki-страницам."],
	["left_menu",256,"Добавление ссылки на приложение в меню слева."],
	["status",1024,"Доступ к статусу пользователя."],
	["offers",32,"Доступ к предложениям (устаревшие методы)."],
	["questions",64,"Доступ к вопросам (устаревшие методы)."],
	["wall",8192,"Доступ к обычным и расширенным методам работы со стеной. "],
	["groups",262144,"Доступ к группам пользователя."],
	["messages",4096,"Доступ к расширенным методам работы с сообщениями."],
	["email",4194304,"Доступ к email пользователя."],
	["notifications",524288,"Доступ к оповещениям об ответах пользователю."],
	["stats",1048576,"Доступ к статистике групп и приложений пользователя, администратором которых он является."],
	["ads",32768,"Доступ к расширенным методам работы с рекламным API."],
	["market",134217728,"Доступ к товарам."],
	["offline",65536,"Доступ к API в любое время"]
],
methods:[
	"users.get","users.search","users.isAppUser","users.getSubscriptions","users.getFollowers","users.report","users.getNearby","auth.checkPhone","auth.signup","auth.confirm","auth.restore","wall.get","wall.search","wall.getById","wall.post","wall.repost","wall.getReposts","wall.edit","wall.delete","wall.restore","wall.pin","wall.unpin","wall.getComments","wall.addComment","wall.editComment","wall.deleteComment","wall.restoreComment","wall.reportPost","wall.reportComment","photos.createAlbum","photos.editAlbum","photos.getAlbums","photos.get","photos.getAlbumsCount","photos.getById","photos.getUploadServer","photos.getOwnerPhotoUploadServer","photos.getChatUploadServer","photos.getMarketUploadServer","photos.getMarketAlbumUploadServer","photos.saveMarketPhoto","photos.saveMarketAlbumPhoto","photos.saveOwnerPhoto","photos.saveWallPhoto","photos.getWallUploadServer","photos.getMessagesUploadServer","photos.saveMessagesPhoto","photos.report","photos.reportComment","photos.search","photos.save","photos.copy","photos.edit","photos.move","photos.makeCover","photos.reorderAlbums","photos.reorderPhotos","photos.getAll","photos.getUserPhotos","photos.deleteAlbum","photos.delete","photos.restore","photos.confirmTag","photos.getComments","photos.getAllComments","photos.createComment","photos.deleteComment","photos.restoreComment","photos.editComment","photos.getTags","photos.putTag","photos.removeTag","photos.getNewTags","friends.get","friends.getOnline","friends.getMutual","friends.getRecent","friends.getRequests","friends.add","friends.edit","friends.delete","friends.getLists","friends.addList","friends.editList","friends.deleteList","friends.getAppUsers","friends.getByPhones","friends.deleteAllRequests","friends.getSuggestions","friends.areFriends","friends.getAvailableForCall","friends.search","widgets.getComments","widgets.getPages","storage.get","storage.set","storage.getKeys","status.get","status.set","audio.get","audio.getById","audio.getLyrics","audio.search","audio.getUploadServer","audio.save","audio.add","audio.delete","audio.edit","audio.reorder","audio.restore","audio.getAlbums","audio.addAlbum","audio.editAlbum","audio.deleteAlbum","audio.moveToAlbum","audio.setBroadcast","audio.getBroadcastList","audio.getRecommendations","audio.getPopular","audio.getCount","pages.get","pages.save","pages.saveAccess","pages.getHistory","pages.getTitles","pages.getVersion","pages.parseWiki","pages.clearCache","groups.isMember","groups.getById","groups.get","groups.getMembers","groups.join","groups.leave","groups.search","groups.getInvites","groups.getInvitedUsers","groups.banUser","groups.unbanUser","groups.getBanned","groups.create","groups.edit","groups.editplace","groups.getSettings","groups.getRequests","groups.editManager","groups.invite","groups.addLink","groups.deleteLink","groups.editLink","groups.reorderLink","groups.removeUser","groups.approveRequest","board.getTopics","board.getComments","board.addTopic","board.addComment","board.deleteTopic","board.editTopic","board.editComment","board.restoreComment","board.deleteComment","board.openTopic","board.closeTopic","board.fixTopic","board.unfixTopic","video.get","video.edit","video.add","video.save","video.delete","video.restore","video.search","video.getUserVideos","video.getAlbums","video.getAlbumById","video.addAlbum","video.editAlbum","video.deleteAlbum","video.reorderAlbums","video.reorderVideos","video.addToAlbum","video.removeFromAlbum","video.getAlbumsByVideo","video.getComments","video.createComment","video.deleteComment","video.restoreComment","video.editComment","video.getTags","video.putTag","video.removeTag","video.getNewTags","video.report","video.reportComment","notes.get","notes.getById","notes.add","notes.edit","notes.delete","notes.getComments","notes.createComment","notes.editComment","notes.deleteComment","notes.restoreComment","places.add","places.getById","places.search","places.checkin","places.getCheckins","places.getTypes","account.getCounters","account.setNameInMenu","account.setOnline","account.setOffline","account.lookupContacts","account.registerDevice","account.unregisterDevice","account.setSilenceMode","account.getPushSettings","account.setPushSettings","account.getAppPermissions","account.getActiveOffers","account.banUser","account.unbanUser","account.getBanned","account.getInfo","account.setInfo","account.changePassword","account.getProfileInfo","account.saveProfileInfo","messages.get","messages.getDialogs","messages.getById","messages.search","messages.getHistory","messages.send","messages.delete","messages.deleteDialog","messages.restore","messages.markAsRead","messages.markAsImportant","messages.getLongPollServer","messages.getLongPollHistory","messages.getChat","messages.createChat","messages.editChat","messages.getChatUsers","messages.setActivity","messages.searchDialogs","messages.addChatUser","messages.removeChatUser","messages.getLastActivity","messages.setChatPhoto","messages.deleteChatPhoto","newsfeed.get","newsfeed.getRecommended","newsfeed.getComments","newsfeed.getMentions","newsfeed.getBanned","newsfeed.addBan","newsfeed.deleteBan","newsfeed.ignoreItem","newsfeed.unignoreItem","newsfeed.search","newsfeed.getLists","newsfeed.saveList","newsfeed.deleteList","newsfeed.unsubscribe","newsfeed.getSuggestedSources","likes.getList","likes.add","likes.delete","likes.isLiked","polls.getById","polls.addVote","polls.deleteVote","polls.getVoters","polls.create","polls.edit","docs.get","docs.getById","docs.getUploadServer","docs.getWallUploadServer","docs.save","docs.delete","docs.add","fave.getUsers","fave.getPhotos","fave.getPosts","fave.getVideos","fave.getLinks","fave.addUser","fave.removeUser","fave.addGroup","fave.removeGroup","fave.addLink","fave.removeLink","notifications.get","notifications.markAsViewed","stats.get","stats.trackVisitor","stats.getPostReach","search.getHints","apps.getCatalog","apps.get","apps.sendRequest","apps.deleteAppRequests","apps.getFriendsList","apps.getLeaderboard","apps.getScore","utils.checkLink","utils.resolveScreenName","utils.getServerTime","database.getCountries","database.getRegions","database.getStreetsById","database.getCountriesById","database.getCities","database.getCitiesById","database.getUniversities","database.getSchools","database.getSchoolClasses","database.getFaculties","database.getChairs","gifts.get","market.get","market.getById","market.search","market.getAlbums","market.getAlbumById","market.createComment","market.getComments","market.deleteComment","market.restoreComment","market.editComment","market.reportComment","market.report","market.add","market.edit","market.delete","market.restore","market.reorderItems","market.reorderAlbums","market.addAlbum","market.editAlbum","market.deleteAlbum","market.removeFromAlbum","market.addToAlbum","store.getStockItems","store.getProducts","captcha.force","chronicle.getFeed","chronicle.search","chronicle.getProfile","chronicle.getPhotos","chronicle.getBanned","chronicle.getPlaces","chronicle.getMessages","chronicle.getUploadServer","chronicle.save","chronicle.getMessageById","chronicle.getFeedback","chronicle.getPreset","chronicle.getExplore","chronicle.addPreset","chronicle.getPhotosById","chronicle.getExploreSection","chronicle.markAsRead","chronicle.getPresets","chronicle.banNews","chronicle.unbanNews","groups.getCatalog","groups.getCatalogInfo","groups.editPlace","video.getCatalog","video.getCatalogSection","video.hideCatalogSection","messages.getHistoryAttachments","docs.getTypes","docs.search","docs.edit","fave.getMarketItems","market.getCategories","ads.getUploadURL"
	],
cart:{
		default_place:1,
		list:[],
		pause:0,
		timer:setInterval(function(){
			if(vk.cart.list.length > 0){
				if(vk.cart.pause == 0){
					vk.cart.list[0][1][3] = 0;
					vk.cart.list[0][0].apply(null,vk.cart.list[0][1]);
					vk.cart.list.shift();
				}
			}
		},335),
	},
_api:function(method,data,callback,place){
	//если нету callback'а
	if(typeof data !== "object")data = {};
	if(typeof callback !== "function")callback = vk.default_callback;
	if(vk.cart && typeof place == "undefined")place = vk.cart.default_place;
	//очередь
	if(!vk.cart || !place){//если не требуется очередь
	}else if(place == -1){//в начало очереди
		return vk.cart.list.unshift([vk._api,[method,data,callback]]);
	}else if(place == 1){//в конец очереди
		return vk.cart.list.push([vk._api,[method,data,callback]]);
	}
	//сбор url
	var url = vk.api_url + method + "?random=" + Math.random();
	if(vk.access_token && !data["access_token"])data["access_token"] = vk.access_token;
	if(!data["v"])data["v"] = vk.default_version;
	//for(var a in data)data[a] = encodeURIComponent(data[a]);
	vk.default_method(url,data,function(body){
		if(typeof body.response !== "undefined"){
			callback(body);
		}else if(body.error){
			if(data["error"] === 1 || (typeof data["error"] == "object" && data["error"].indexOf(body.error.error_code) > -1))callback(body);
			else vk.on.error(body.error,method,data,callback);
		}
	},function(){
		if(data["error"])callback({error:{error_code:404,error_msg:"Нет соединения с интернетом"}});
		else vk.on.error({error_code:404,error_msg:"Нет соединения с интернетом"},method,data,callback);
	});
},
_apiSync:function(method,data){
	if(vk.inbrowser)return console.error("The synchronous method is disabled in your browser");
	if(typeof data !== "object")data = {};//если нету data
	//сбор url
	var url = vk.api_url + method + "?random=" + Math.random();
	if(vk.access_token && !data["access_token"])data["access_token"] = vk.access_token;
	if(!data["v"])data["v"] = vk.default_version;
	//for(var a in data)data[a] = encodeURIComponent(data[a]);
	return vk.sreq.postSync(url,data);
},
addListener:{
	messages:function(_cb,data){
		if(!data)data = {};
		if(!data.access_token)data.access_token = vk.access_token;
		if(!_cb)return false;
		var pts,last_msg_id,users_info = {};
		if(vk.debag)console.log("Получаю сервер...");
		vk.api.messages.getLongPollServer({use_ssl:0,need_pts:1,access_token:data.access_token},function(res){
			pts = res.response.pts;
			if(vk.debag)console.log("Сервер успешно получен.");
			setInterval(function(){
				vk.api.messages.getLongPollHistory({pts:pts,access_token:data.access_token,fields:data.fields},function(res){
					if(vk.debag)console.log("Сервер успешно ответил...");
					if( !pts || pts < res.response.new_pts)pts = res.response.new_pts;
					delete res.response.profiles.map(function(profile){users_info[profile.id] = profile;});
					delete res.response.messages.items.map(function(msg){
						if(typeof msg == "number" || last_msg_id >= msg.id)return;
							last_msg_id = msg.id;
							msg.users_info = users_info;
							msg.user_info = (users_info[msg.user_id] || {});
							msg.peer_id = (msg.chat_id?msg.chat_id + 2000000000:msg.user_id);
							msg.reply = function(text,other,__cb,place){
								if(typeof other !== "object")other = {}
								other.peer_id = msg.peer_id;
								other.forward_messages = msg.id;
								other.access_token = data.access_token;
								other.message = ((typeof text == "string")?text:JSON.stringify(text));
								vk.api.messages.send(other,__cb || function(){},place || 0);
							};
							msg.send = function(text,other,__cb,place){
								if(typeof other !== "object")other = {}
								other.peer_id = msg.peer_id;
								other.access_token = data.access_token;
								other.message = ((typeof text == "string")?text:JSON.stringify(text));
								vk.api.messages.send(other,__cb || function(){},place || 0);
							};
							_cb(msg);
						});
				});
			},data.interval || 500);
		});
	}
},
on:{
	captcha:function(_error,_m,_q,_cb){
		if(vk.debag)console.info("Был запрошен защитный код, игнорируем...",_error);
		},
	captchaSubmit:function(v){
		vk.cart.list[0][1][1].captcha_key = v;
		vk.cart.pause = 0;
		},
	error:function(_error,_m,_q,_cb){
			if(_error.error_code == 6){//Too many requests per second
				vk._api(_m,_q,_cb,-1);
			}else if(_error.error_code == 14){//captcha
				vk.on.captcha(_error,_m,_q,_cb);
			}else{
				vk.on.otherError(_error,_m,_q,_cb);
			}
		},
	otherError:function(_error,_m,_q,_cb){
		if(vk.debag)console.log("on.error",arguments);
	}
}
}

vk.default_method = browser?vk.sreq.jsonp:vk.sreq.post;//стандартный метод для запросов
//заполняем секции vk.api и vk.apiSync методами
vk.apiSync = {};
vk.api = {};

for(var m in vk.methods){
	m = vk.methods[m];
	if(!vk.api[m.split(".")[0]])vk.api[m.split(".")[0]] = {};
	if(!vk.apiSync[m.split(".")[0]])vk.apiSync[m.split(".")[0]] = {};
	vk.api[m.split(".")[0]][m.split(".")[1]] = (function(method){
		return function(data,callback,place){
			return vk._api(method,data,callback,place);
		}
	}
	)(m);
	vk.apiSync[m.split(".")[0]][m.split(".")[1]] = (function(method){
		return function(data){
			return vk._apiSync(method,data);
		}
	}
	)(m);
}

return vk;
})();

if(typeof module == "undefined"){
	var CallbackRegistry = {}; // реестр для jsonp
}else{
	module.exports = vk;
}
